### PR TITLE
Add path numeration for derived accounts

### LINKED ISSUE #73: Default derivation path calculation
//0 for first child
//1 for second
etc.

### PR DESCRIPTION
**Pull request by VladStarostenko**
close #73

### CODE REVIEW

File: packages/extension-ui/src/partials/DerivationPath.tsx [hunk 1]
```diff
@@ -11,11 +11,13 @@ interface Props {
...
   parentPassword: string;
+  siblingCount: number;
 }
 
-function DerivationPath ({ onChange, parentAddress, parentPassword }: Props): React.ReactElement<Props> {
-  const [path, setPath] = useState<string>('');
+function DerivationPath ({ onChange, parentAddress, parentPassword, siblingCount }: Props): React.ReactElement<Props> {
+  const [path, setPath] = useState<string>(`//${siblingCount}`);
 
+  console.log(path);
```
Comment (krzysztof-jelski) [inline packages/extension-ui/src/partials/DerivationPath.tsx]: Remove `console.log`

File: packages/extension-ui/src/components/ValidatedInput.tsx [hunk 2]
```diff
@@ -17,18 +17,21 @@ type Props<T extends BasicProps> = T & {
...
+  const [value, setValue] = useState(defaultValue || '');
   const [wasMounted, setWasMounted] = useState(false);
   const [validationResult, setValidationResult] = useState<Result<string>>(Result.ok(''));
 
   useEffect(() => {
     if (!wasMounted) {
       setWasMounted(true);
 
-      return;
+      if (!defaultValue) {
```
Comment (krzysztof-jelski) [inline packages/extension-ui/src/components/ValidatedInput.tsx]: We don't validate if there is no default value? Perhaps worth a test for better explanation.

File: packages/extension-ui/src/Popup/Derive/index.tsx [hunk 3]
```diff
@@ -80,6 +79,7 @@ function Derive ({ match: { params: { address: parentAddress } } }: Props): Reac
         onChange={setAccount}
         parentAddress={parentAddress}
         parentPassword={parentPassword}
+        siblingCount={childrenCount(accounts, parentAddress)}
```
Comment (krzysztof-jelski) [inline packages/extension-ui/src/Popup/Derive/index.tsx]: That would be `siblingsCount` (plural).

Still I'm a bit unsure about the naming here. We have a bit of "business" logic here: the next derivation path should be equal to current count number of siblings.

The question is where should it go? Here, the DerivationPath, or perhaps a new file, that will include more complex rules for creating derivation paths in the future?

So perhaps here we should have `defaultDerivationPath={nextDerivationPath(childrenCount(accounts, parentAddress))}`

File: packages/extension-ui/src/partials/DerivationPath.tsx [hunk 4]
```diff
@@ -34,6 +36,7 @@ function DerivationPath ({ onChange, parentAddress, parentPassword }: Props): Re
   return (
     <ValidatedInput
       component={InputWithLabel}
+      defaultValue={`//${siblingCount}`}
```
Comment (krzysztof-jelski) [inline packages/extension-ui/src/partials/DerivationPath.tsx]: The logic of creating the next derivation path is duplicated (l. 18)

**Review by krzysztof-jelski — CHANGES_REQUESTED** [review]
left some comments

File: packages/extension-ui/src/utils/nextDerivationPath.test.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  test('generates path for first masters child', () => {
+    testHierarchy([acc('a')], 'a', '//0');
+  });
+
+  test('generates path for third masters child', () => {
+    testHierarchy([acc('a'), acc('b', 'a'), acc('c', 'a')], 'a', '//2');
+  });
+
+  test('generates path for masters child when exist one more root', () => {
```
Comment (vanruch) [inline packages/extension-ui/src/utils/nextDerivationPath.test.ts]:
```suggestion
  test('generates path for master's child when another root exists', () => {
```

**Review by vanruch — APPROVED** [review]

File: packages/extension-ui/src/utils/nextDerivationPath.test.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2019-2020 @polkadot/extension-ui authors & contributors
+// This software may be modified and distributed under the terms
+// of the Apache-2.0 license. See the LICENSE file for details.
+
+import { AccountJson } from '@polkadot/extension-base/background/types';
+import { nextDerivationPath } from './nextDerivationPath';
+
+const testHierarchy = (accounts: AccountJson[], parentAddress: string, expected: string): void => {
```
Comment (krzysztof-jelski) [inline packages/extension-ui/src/utils/nextDerivationPath.test.ts]: I believe the tests would be more readable without this method extracted. I suggest inlining it.

**Review by krzysztof-jelski — APPROVED** [review]

**Pull request merged**